### PR TITLE
MINIFICPP-1117 - minifi::Exception is now nothrow copyable

### PR DESCRIPTION
**Pull request by szaszm**
Implemented in terms of std::runtime_error

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by szaszm** [comment]
I've added some boilerplate for string joining. I can remove it at some performance cost (extra `std::string` allocations) or additional dependency (`boost::algorithm::join`)

File: libminifi/include/Exception.h [hunk 1]
```diff
@@ -44,49 +43,56 @@ enum ExceptionType {
...
 
- private:
-  // Exception type
-  ExceptionType _type;
-  // Exception detailed information
-  std::string _errorMsg;
-  // Hold the what result
-  mutable std::string _whatStr;
+template<typename... Strs>
+std::string StringJoin(Strs&&... strs) {
```
Comment (nghiaxlee) [inline libminifi/include/Exception.h]: I think this should be moved to our StringUtils if we haven't had this feature yet.

Comment (arpadboda) [reply libminifi/include/Exception.h]: Good point, we do have a join there.

File: libminifi/include/Exception.h [hunk 2]
```diff
@@ -44,49 +43,56 @@ enum ExceptionType {
...
-  }
+namespace detail {
+inline size_t StringLength(const char* str) { return strlen(str); }
 
-  // Destructor
-  virtual ~Exception() noexcept {
-  }
-  virtual const char * what() const noexcept {
+template<size_t L>
+constexpr size_t StringLength(const char (&str)[L]) { return L; }
```
Comment (szaszm) [inline libminifi/include/Exception.h]: bug here: L is the size of the array, not the string inside. I'll change it to `max(L-1, 0)` and leave the case of early null termination unhandled if the functionality is not already present in StringUtils.

File: libminifi/include/Exception.h (hunk 1 above)
Comment (szaszm) [reply libminifi/include/Exception.h]: moved to StringUtils in dbc0033

**Comment by szaszm** [comment]
update:
- fix `ProcessorTests`
- move string utilities to `StringUtils` and generalize them
- add unit test to cover the new `StringUtils::join_pack`
Moving the utilities from a private namespace to `StringUtils` means that they need to be more general IMO. I've extended them to work with any character type and do SFINAE checking on the string types. 
This means quite some template code, which may not be welcome in this codebase. Suggestions are welcome.

**Pull request closed without merging**